src/com/google/cose/Ec2SigningKey.java

```diff
@@ -0,0 +1,222 @@
+        cborKey.put(new NegativeInteger(Headers.KEY_PARAMETER_X), new ByteString(xCor));
+      }
+      if (yCor != null) {
+        cborKey.put(new NegativeInteger(Headers.KEY_PARAMETER_Y), new ByteString(yCor));
+      }
+      if (dCor != null) {
+        cborKey.put(new NegativeInteger(Headers.KEY_PARAMETER_D), new ByteString(dCor));
+      }
+      return new Ec2SigningKey(cborKey);
+    }
+
+    public Builder withKeyId(String keyId) {
+      this.keyId = keyId;
+      return this;
+    }
+
+    public Builder withAlgorithm(Algorithm algorithm) {
+      this.algorithm = algorithm;
+      return this;
+    }
+
+    public Builder withOperations(Integer...operations) {
+      this.operations.addAll(Arrays.asList(operations));
+      return this;
+    }
+
+    public Builder withBaseIv(byte[] baseIv) {
+      this.baseIv = baseIv;
+      return this;
+    }
+
+    public Builder withCurve(int curve) throws CoseException {
+      if ((curve < 0) || (curve > Headers.CURVE_EC2_P521)) {
+        throw new CoseException("Unsupported Curve provided.");
+      }
+      this.curve = curve;
+      return this;
+    }
+
+    public Builder withXCoordinate(byte[] xCor) {
+      this.xCor = xCor;
+      return this;
+    }
+
+    public Builder withYCoordinate(byte[] yCor) {
+      this.yCor = yCor;
+      return this;
+    }
+
+    public Builder withDCoordinate(byte[] dCor) {
+      this.dCor = dCor;
+      return this;
+    }
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  private KeyPair getKeyPairFromCbor() throws CoseException, CborException {
+    if (keyType != Headers.KEY_TYPE_EC2) {
+      throw new CoseException("Expecting EC2 key (type 2), found type " + keyType);
+    }
+
+    final KeyPair keyPair;
+    // Get curve information
+    int curve = CborUtils.asInteger(labels.get(Headers.KEY_PARAMETER_CURVE));
+
+    // Get private key.
+    final PrivateKey privateKey;
+    if (labels.containsKey(Headers.KEY_PARAMETER_D)) {
+      final ByteString key = CborUtils.asByteString(labels.get(Headers.KEY_PARAMETER_D));
+      privateKey = CoseUtils.getEc2PrivateKeyFromCoordinate(
+          curve,
+          new BigInteger(SIGN_POSITIVE, key.getBytes())
+      );
+    } else {
+      privateKey = null;
+    }
+
+    if (!labels.containsKey(Headers.KEY_PARAMETER_X)) {
+      if (privateKey == null) {
+        throw new IllegalStateException("Missing key material information.");
+      } else {
+        return new KeyPair(null, privateKey);
+      }
+    }
+
+    final ByteString xCor = CborUtils.asByteString(labels.get(Headers.KEY_PARAMETER_X));
+    // Get the public key for EC2 key.
+    // We should not have a case where x is provided but y is not.
+    if (!labels.containsKey(Headers.KEY_PARAMETER_Y)) {
+      throw new IllegalStateException("X coordinate provided but Y coordinate is missing.");
+    }
+    final ByteString yCor = CborUtils.asByteString(labels.get(Headers.KEY_PARAMETER_Y));
```

Comment:
We should support compressed format. Seems like we only support full coordinates. From the RFC:

```
   +-------+------+-------+--------+-----------------------------------+
   | Key   | Name | Label | CBOR   | Description                       |
   | Type  |      |       | Type   |                                   |
   +-------+------+-------+--------+-----------------------------------+
   | 2     | crv  | -1    | int /  | EC identifier - Taken from the    |
   |       |      |       | tstr   | "COSE Elliptic Curves" registry   |
   | 2     | x    | -2    | bstr   | x-coordinate                      |
   | 2     | y    | -3    | bstr / | y-coordinate                      |
   |       |      |       | bool   |                                   |
   | 2     | d    | -4    | bstr   | Private key                       |
   +-------+------+-------+--------+-----------------------------------+
```

Feel free to file a bug to track this as a feature request.

Reply:
Added an issue for this.